extensions/sql/processors/ExecuteSQL.cpp

```diff
@@ -0,0 +1,150 @@
+		"The maximum number of result rows that will be included intoi a flow file. If zero then all will be placed into the flow file")->supportsExpressionLanguage(true)->build());
+
+const core::Relationship ExecuteSQL::s_success("success", "Successfully created FlowFile from SQL query result set.");
+
+ExecuteSQL::ExecuteSQL(const std::string& name, utils::Identifier uuid)
+    : core::Processor(name, uuid), max_rows_(0),
+      logger_(logging::LoggerFactory<ExecuteSQL>::getLogger()) {
+}
+
+ExecuteSQL::~ExecuteSQL() {
+}
+
+void ExecuteSQL::initialize() {
+  //! Set the supported properties
+  setSupportedProperties( { s_dbControllerService, s_sqlSelectQuery, s_maxRowsPerFlowFile });
+
+  //! Set the supported relationships
+  setSupportedRelationships( { s_success });
+}
+
+void ExecuteSQL::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  context->getProperty(s_dbControllerService.getName(), db_controller_service_);
+  context->getProperty(s_sqlSelectQuery.getName(), sqlSelectQuery_);
+  context->getProperty(s_maxRowsPerFlowFile.getName(), max_rows_);
+
+  database_service_ = std::dynamic_pointer_cast<sql::controllers::DatabaseService>(context->getControllerService(db_controller_service_));
+  if (!database_service_)
+    throw minifi::Exception(PROCESSOR_EXCEPTION, "'DB Controller Service' must be defined");
+}
+
+void ExecuteSQL::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  std::unique_lock<std::mutex> lock(onTriggerMutex_, std::try_to_lock);
+  if (!lock.owns_lock()) {
+    logger_->log_warn("'onTrigger' is called before previous 'onTrigger' call is finished.");
+    context->yield();
+    return;
+  }
+
+  if (!connection_) {
+    connection_ = database_service_->getConnection();
+    if (!connection_) {
+      context->yield();
+      return;
+    }
+  }
+
+  try {
+    auto statement = connection_->prepareStatement(sqlSelectQuery_);
+
+    auto rowset = statement->execute();
+
+    int count = 0;
+    size_t row_count = 0;
+    std::stringstream outputStream;
+    sql::JSONSQLWriter writer(rowset, &outputStream);
+    // serialize the rows
+    do {
+      row_count = writer.serialize(max_rows_ == 0 ? std::numeric_limits<size_t>::max() : max_rows_);
+      count++;
+      if (row_count == 0)
+        break;
+      writer.write();
+      auto output = outputStream.str();
+      if (!output.empty()) {
+        WriteCallback writer(output.data(), output.size());
+        auto newflow = session->create();
+        newflow->addAttribute("executesql.resultset.index", std::to_string(row_count));
+        session->write(newflow, &writer);
+        session->transfer(newflow, s_success);
+      }
+      outputStream.str("");
+      outputStream.clear();
+    } while (row_count > 0);
+  } catch (std::exception& e) {
+    logger_->log_error(e.what());
```

Comment:
(!) This seems to be swallowing the exception. 
It's good that we log the error, although in case an error occurs we leave some flow files in the session already transferred. 

Not sure what would be the correct behaviour here, but I feel that this is definitely not one of those.

We can either throw the exception to trigger rollback or we could introduce error handling at flowfile level and route to failure in case something went wrong. 

For the ease of simplicity I would just throw the exception again to make sure the session is rolled back.

Reply:
Fixed.